File: python/src/graph_engineering/adapters/retry.py
"""Retry decisions, rules ``R-001`` through ``R-006``.

adapter-semantics 8.4: backoff is integer arithmetic with no jitter, because
runtime-capability-semantics admits retry without jitter only and a jittered
schedule is not reproducible across two runtimes::

    computed(n)  = min(maxBackoffMs, floor(initialBackoffMs * m^(n-1) / 1000^(n-1)))
    effective(n) = retryAfterMs is None ? computed(n) : max(computed(n), retryAfterMs)

A provider hint above the local ceiling abandons the attempt rather than
truncating it: truncating would re-dispatch inside exactly the window the
provider refused.
"""

from __future__ import annotations

from .contract import SIDE_EFFECT_ORDER, side_effect_permits_retry, taxonomy_facts
from .types import AdapterDescriptor, AdapterRetryPolicy, RetryDecision


def computed_backoff_ms(policy: AdapterRetryPolicy, attempt: int) -> int:
    """Exact integer arithmetic.  Python integers keep a long schedule portable."""
    numerator = policy.initial_backoff_ms
    denominator = 1
    for _ in range(1, attempt):
        numerator *= policy.backoff_multiplier_milli
        denominator *= 1000
    value = numerator // denominator
    return min(value, policy.max_backoff_ms)


def retry_decision(
    descriptor: AdapterDescriptor,
    code: str,
    side_effect_class: str,
    attempt: int,
    retry_after_ms: int | None,
) -> RetryDecision:
    facts = taxonomy_facts(code)
    if side_effect_class not in SIDE_EFFECT_ORDER:
        raise ValueError(f"unknown side-effect class {side_effect_class!r}")
    if attempt < 1:
        raise ValueError("attempt must be a positive integer")
    if retry_after_ms is not None and retry_after_ms < 0:
        raise ValueError("retryAfterMs must be None or a non-negative integer")
    policy = descriptor.retry_policy

    if not facts.retryable:
        return RetryDecision(may_retry=False, backoff_ms=None, rule="R-005")
    if attempt >= policy.max_attempts:
        return RetryDecision(may_retry=False, backoff_ms=None, rule="R-004")
    if not side_effect_permits_retry(facts.effect_disposition, side_effect_class):
        return RetryDecision(may_retry=False, backoff_ms=None, rule="R-006")
    computed = computed_backoff_ms(policy, attempt)
    if retry_after_ms is None:
        return RetryDecision(may_retry=True, backoff_ms=computed, rule="R-001")
    effective = max(computed, retry_after_ms)
    if effective > policy.max_backoff_ms:
        return RetryDecision(may_retry=False, backoff_ms=None, rule="R-003")
    return RetryDecision(may_retry=True, backoff_ms=effective, rule="R-002")

File: packages/adapters/src/retry.ts
/**
 * Retry decisions, rules `R-001` through `R-006`.
 *
 * adapter-semantics 8.4: backoff is integer arithmetic with no jitter, because
 * runtime-capability-semantics admits retry without jitter only and a jittered
 * schedule is not reproducible across two runtimes.
 *
 *   computed(n)  = min(maxBackoffMs, floor(initialBackoffMs * m^(n-1) / 1000^(n-1)))
 *   effective(n) = retryAfterMs == null ? computed(n) : max(computed(n), retryAfterMs)
 *
 * A provider hint above the local ceiling abandons the attempt rather than
 * truncating it: truncating would re-dispatch inside exactly the window the
 * provider refused.
 */

import { SIDE_EFFECT_ORDER, sideEffectPermitsRetry, taxonomyFacts } from "./contract.js";
import type {
  AdapterDescriptor,
  AdapterErrorCode,
  AdapterRetryPolicy,
  RetryDecision,
  SideEffectClass,
} from "./types.js";

/** Exact integer arithmetic; `BigInt` keeps a long schedule portable. */
export function computedBackoffMs(policy: AdapterRetryPolicy, attempt: number): number {
  let numerator = BigInt(policy.initialBackoffMs);
  let denominator = 1n;
  for (let step = 1; step < attempt; step += 1) {
    numerator *= BigInt(policy.backoffMultiplierMilli);
    denominator *= 1000n;
  }
  const value = numerator / denominator;
  const ceiling = BigInt(policy.maxBackoffMs);
  return Number(value > ceiling ? ceiling : value);
}

export function retryDecision(
  descriptor: AdapterDescriptor,
  code: AdapterErrorCode,
  sideEffectClass: SideEffectClass,
  attempt: number,
  retryAfterMs: number | null,
): RetryDecision {
  const facts = taxonomyFacts(code);
  if (!SIDE_EFFECT_ORDER.includes(sideEffectClass)) {
    throw new Error(`unknown side-effect class '${String(sideEffectClass)}'`);
  }
  if (!Number.isSafeInteger(attempt) || attempt < 1) {
    throw new Error("attempt must be a positive integer");
  }
  if (retryAfterMs !== null && (!Number.isSafeInteger(retryAfterMs) || retryAfterMs < 0)) {
    throw new Error("retryAfterMs must be null or a non-negative integer");
  }
  const policy = descriptor.retryPolicy;

  if (!facts.retryable) {
    return Object.freeze({ mayRetry: false, backoffMs: null, rule: "R-005" as const });
  }
  if (attempt >= policy.maxAttempts) {
    return Object.freeze({ mayRetry: false, backoffMs: null, rule: "R-004" as const });
  }
  if (!sideEffectPermitsRetry(facts.effectDisposition, sideEffectClass)) {
    return Object.freeze({ mayRetry: false, backoffMs: null, rule: "R-006" as const });
  }
  const computed = computedBackoffMs(policy, attempt);
  if (retryAfterMs === null) {
    return Object.freeze({ mayRetry: true, backoffMs: computed, rule: "R-001" as const });
  }
  const effective = Math.max(computed, retryAfterMs);
  if (effective > policy.maxBackoffMs) {
    return Object.freeze({ mayRetry: false, backoffMs: null, rule: "R-003" as const });
  }
  return Object.freeze({ mayRetry: true, backoffMs: effective, rule: "R-002" as const });
}
